### PR TITLE
fix: #CRRE-563 consider articles without credits in price calculation

### PR DESCRIPTION
**Pull request by BastienBenard**
## Describe your changes
On the validation page, articles that come from campaigns without credits cause errors.

## Checklist tests
As a validator, I have several baskets with at least one basket coming from a campaign without credits (item A).

When I select only item A, the quantity of selected items increases, but the other variables remain unchanged.
If I select all items and then deselect item A, only the quantity changes.
If I select all items and then deselect another item, the values decrement correctly based on the values of the deselected item.
Make sure that you only have to validate article A.
When I select it, the quantity of selected items increases, but the other variables remain unchanged.

Test exports.

## Issue ticket number and link
[CRRE-563](https://jira.support-ent.fr/browse/CRRE-563)

## Checklist before requesting a review (magic string, indentation, comment/documentation...)

- [x] I have detailed the tests to do in my feature/fix in order to prevent consequents regressions (must specify in **Checklist tests**)
- [x] I have performed a self-review of my code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation (API Doc etc...) - (must specify in **Description** for target version)
- [ ] If it is a consequent feature, I have added thorough tests.
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] Any dependent changes have been added to this project (must specify in **Description**)

### CODE REVIEW

File: src/main/resources/public/ts/model/OrderSearchAmount.ts [hunk 1]
```diff
@@ -19,13 +19,13 @@ export class OrderSearchAmount implements IOrderSearchAmount{
...
-        this._total = data.total;
-        this._total_filtered = data.total_filtered;
-        this._total_filtered_consumable = data.total_filtered_consumable;
+        this._licence = data.licence || 0;
+        this._consumable_licence = data.consumable_licence || 0;
+        this._credit = data.credit || 0;
+        this._consumable_credit = data.consumable_credit || 0;
+        this._total = data.total || 0;
+        this._total_filtered = data.total_filtered || 0;
+        this._total_filtered_consumable = data.total_filtered_consumable || 0;
```
Comment (BastienBenard) [inline src/main/resources/public/ts/model/OrderSearchAmount.ts:28]: Permet d'éviter les "undefined".

File: src/main/resources/public/ts/controllers/validator/orders-waiting.ts [hunk 2]
```diff
@@ -54,7 +62,15 @@ export const waitingValidatorOrderController = ng.controller('waitingValidatorOr
...
-            $scope.amountTotal = await $scope.ordersClient.calculateTotal('WAITING', $scope.current.structure.id, $scope.filterOrder);
+            let filter: OrderSearchAmountFilter = new OrderSearchAmountFilter();
+            filter.status = [ORDER_STATUS_ENUM.WAITING];
+            filter.idsCampaign = $scope.filterOrder.typeCampaignList.map((campaign: Campaign) => campaign.id);
+            filter.idsUser = $scope.filterOrder.userList.map((user: UserModel) => user.id_user);
+            filter.searchingText = $scope.filterOrder.queryName;
+            let {startDate, endDate} = Utils.formatDate($scope.filterOrder.startDate, $scope.filterOrder.endDate);
+            filter.startDate = startDate;
+            filter.endDate = endDate;
+            $scope.amountTotal = await $scope.ordersClient.calculateTotal($scope.current.structure.id, filter);
```
Comment (Sofianebernoussi) [inline src/main/resources/public/ts/controllers/validator/orders-waiting.ts:73]: y'a pas un champ list structure dans filter?

**Review by Sofianebernoussi — APPROVED** [review]

File: src/main/java/fr/openent/crre/controllers/OrderController.java [hunk 3]
```diff
@@ -151,132 +151,53 @@ public void listOrders(final HttpServerRequest request) {
...
+            OrderHelper.listOrderAndCalculatePrice(filters, request)
+                    .onSuccess(orderUniversalModels -> {
+                        int nbItem = 0;
+                        int nbLicence = 0;
+                        int nbConsumableLicence = 0;
+                        double priceCredit = 0.0;
+                        double priceConsumableCredit = 0.0;
+                        for (OrderUniversalModel orderUniversalModel : orderUniversalModels) {
+                            nbItem += orderUniversalModel.getAmount();
+                            switch (CreditTypeEnum.getValue(orderUniversalModel.getCampaign().getUseCredit(), CreditTypeEnum.NONE)) {
```
Comment (alicedraillard) [inline src/main/java/fr/openent/crre/controllers/OrderController.java:171]: default case ?

File: src/main/java/fr/openent/crre/service/OrderService.java [hunk 4]
```diff
@@ -38,13 +38,6 @@ Future<List<OrderUniversalModel>> listOrder(String searchingText, List<Integer>
      */
     void listOrder(String status, String idStructure, Integer page, String startDate, String endDate, Handler<Either<String, JsonArray>> handler);
 
-    void listOrderAmount(String status, String idStructure, UserInfos user, String startDate, String endDate, Boolean consumable,
```
Comment (alicedraillard) [inline src/main/java/fr/openent/crre/service/OrderService.java:41]: explication de pk tu supprimes ca ? --> pas utilisé ? déplacé ?

File: src/main/java/fr/openent/crre/service/impl/DefaultOrderService.java [hunk 5]
```diff
@@ -130,78 +130,6 @@ public void listOrder(String status, String idStructure, Integer page, String st
         sql.prepared(query, values, SqlResult.validResultHandler(handler));
     }
 
-    @Override
```
Comment (alicedraillard) [inline src/main/java/fr/openent/crre/service/impl/DefaultOrderService.java:133]: cf oderService.java

File: src/main/java/fr/openent/crre/security/PrescriberInStructureRight.java [hunk 6]
```diff
@@ -13,10 +13,14 @@ public class PrescriberInStructureRight implements ResourcesProvider {
...
     @Override
     public void authorize(HttpServerRequest httpServerRequest, Binding binding, UserInfos userInfos, Handler<Boolean> handler) {
+        httpServerRequest.pause();
         String idStructure = httpServerRequest.getParam(Field.IDSTRUCTURE) != null ? httpServerRequest.getParam(Field.IDSTRUCTURE) :
                 httpServerRequest.params().get(Field.IDSTRUCTURE);
         WorkflowHelper.authorize(idStructure, userInfos, Crre.PRESCRIPTOR_RIGHT)
-                .onSuccess(handler)
+                .onSuccess(res -> {
+                    handler.handle(res);
+                    httpServerRequest.resume();
```
[the remark this answers is not in the document]
Comment (BastienBenard) [reply src/main/java/fr/openent/crre/security/PrescriberInStructureRight.java:22]: C'est pour éviter les erreurs "request have already read"

File: src/main/java/fr/openent/crre/controllers/OrderController.java (hunk 3 above)
Comment (BastienBenard) [reply src/main/java/fr/openent/crre/controllers/OrderController.java:171]: Pas besoin

File: src/main/java/fr/openent/crre/service/OrderService.java (hunk 4 above)
Comment (BastienBenard) [reply src/main/java/fr/openent/crre/service/OrderService.java:41]: Ce n'est plus utilisé, maintenant on utilise le model OrderUniversalModel

File: src/main/java/fr/openent/crre/service/impl/DefaultOrderService.java (hunk 5 above)
Comment (BastienBenard) [reply src/main/java/fr/openent/crre/service/impl/DefaultOrderService.java:133]: same

File: src/main/resources/public/ts/controllers/validator/orders-waiting.ts (hunk 2 above)
Comment (BastienBenard) [reply src/main/resources/public/ts/controllers/validator/orders-waiting.ts:73]: Je fais cela car je veux explicitement que ce soit une seule structure. Je ne veux pas qu'il soit possible de passer plusieurs structures.

**Review by alicedraillard — APPROVED** [review]

**Pull request merged**